src/Adapter/MSTest.TestAdapter/Extensions/MethodInfoExtensions.cs

```diff
@@ -89,20 +89,17 @@ internal static bool HasCorrectTestMethodSignature(this MethodInfo method, bool
     /// Checks whether test method has correct Timeout attribute.
     /// </summary>
     /// <param name="method">The method to verify.</param>
+    /// <param name="timeoutAttribute">The timeout attribute when we already have it.</param>
     /// <returns>True if the method has the right test timeout signature.</returns>
-    internal static bool HasCorrectTimeout(this MethodInfo method)
+    internal static bool HasCorrectTimeout(this MethodInfo method, TimeoutAttribute? timeoutAttribute = null)
     {
         DebugEx.Assert(method != null, "method should not be null.");
 
-        // There should be one and only one TimeoutAttribute.
-        TimeoutAttribute[] attributes = ReflectHelper.GetCustomAttributes<TimeoutAttribute>(method, false);
-        if (attributes?.Length != 1)
-        {
-            return false;
-        }
+        // TODO: redesign this, probably change this to GetTimeout? so we don't have to do this weird dance?
+        timeoutAttribute ??= ReflectHelper.Instance.GetSingleNonDerivedAttributeOrDefault<TimeoutAttribute>(method, inherit: false, nullOnMultiple: true);
```

Comment:
Or better ideas to redesign? Before the change it was grabbing the attributes 3 times to check: 
- if the attribute is defined
- then again in validation method
- then again to grab the timeout